src/core/EntrySearcher.cpp

```diff
@@ -20,75 +20,163 @@
-    for (const QString& word : wordList) {
-        if (!wordMatch(word, entry, caseSensitivity)) {
-            return QList<Entry*>();
-        }
+    QList<Entry*> results;
+    for (Entry* entry : entries) {
+       if (searchEntryImpl(searchString, entry)) {
+           results.append(entry);
+       }
     }
-
-    return QList<Entry*>() << entry;
+    return results;
 }
 
-bool EntrySearcher::wordMatch(const QString& word, Entry* entry, Qt::CaseSensitivity caseSensitivity)
+void EntrySearcher::setCaseSensitive(bool state)
 {
-    return entry->resolvePlaceholder(entry->title()).contains(word, caseSensitivity) ||
-            entry->resolvePlaceholder(entry->username()).contains(word, caseSensitivity) ||
-            entry->resolvePlaceholder(entry->url()).contains(word, caseSensitivity) ||
-            entry->resolvePlaceholder(entry->notes()).contains(word, caseSensitivity);
+    m_caseSensitive = state;
 }
 
-bool EntrySearcher::matchGroup(const QString& searchTerm, const Group* group, Qt::CaseSensitivity caseSensitivity)
+bool EntrySearcher::searchEntryImpl(const QString& searchString, Entry* entry)
 {
-    const QStringList wordList = searchTerm.split(QRegExp("\\s"), QString::SkipEmptyParts);
-    for (const QString& word : wordList) {
-        if (!wordMatch(word, group, caseSensitivity)) {
+    auto searchTerms = parseSearchTerms(searchString);
+    bool found;
+
+    for (SearchTerm* term : searchTerms) {
+        switch (term->field) {
+        case Field::Title:
+            found = term->regex.match(entry->resolvePlaceholder(entry->title())).hasMatch();
+            break;
+        case Field::Username:
+            found = term->regex.match(entry->resolvePlaceholder(entry->username())).hasMatch();
+            break;
+        case Field::Password:
+            found = term->regex.match(entry->resolvePlaceholder(entry->password())).hasMatch();
+            break;
+        case Field::Url:
+            found = term->regex.match(entry->resolvePlaceholder(entry->url())).hasMatch();
+            break;
+        case Field::Notes:
+            found = term->regex.match(entry->notes()).hasMatch();
+            break;
+        case Field::Attribute:
+            found = !entry->attributes()->keys().filter(term->regex).empty();
+            break;
+        case Field::Attachment:
+            found = !entry->attachments()->keys().filter(term->regex).empty();
+            break;
+        default:
+            found = term->regex.match(entry->resolvePlaceholder(entry->title())).hasMatch() ||
+                    term->regex.match(entry->resolvePlaceholder(entry->username())).hasMatch() ||
+                    term->regex.match(entry->resolvePlaceholder(entry->url())).hasMatch() ||
+                    term->regex.match(entry->notes()).hasMatch();
+        }
+
+        // Short circuit if we failed to match or we matched and are excluding this term
+        if (!found || term->exclude) {
             return false;
```

Comment:
Shouldn't this be a `continue` instead? It is into a for and returning here will skip all next term in the for if the current is excluded

Reply:
It is a short circuit because this function only returns true if all the terms match (implicit AND of terms).

Reply:
ok make sense